### PR TITLE
python312Packages.bitsandbytes: 0.43.1 -> 0.43.3

### PR DESCRIPTION
**Pull request by GaetanLepage**
## Description of changes

Changelog: https://github.com/TimDettmers/bitsandbytes/releases/tag/0.43.3

cc @bcdarwin

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by remexre** [comment]
The CUDA-enabled version of this no longer builds when I do:

```
$ NIXPKGS_ALLOW_UNFREE=1 nix repl --impure
Welcome to Nix 2.18.5. Type :? for help.

nix-repl> :lf .
Added 18 variables.

nix-repl> :b outputs.legacyPackages.x86_64-linux.python3Packages.bitsandbytes.override { torch = outputs.legacyPackages.x86_64-linux.python3Packages.torchWithCuda; }
error: builder for '/nix/store/ylwx6as34nfygmxz5lwjvcqb15p7xd5q-python3.12-bitsandbytes-0.43.3.drv' failed with exit code 1;
       last 10 log lines:
       >
       >
       > Call Stack (most recent call first):
       >   /nix/store/yzi080r2c1zn2jzrhcfdv7dmr92yw07l-cmake-3.29.6/share/cmake-3.29/Modules/CMakeDetermineCompilerId.cmake:8 (CMAKE_DETERMINE_COMPILER_ID_BUILD)
       >   /nix/store/yzi080r2c1zn2jzrhcfdv7dmr92yw07l-cmake-3.29.6/share/cmake-3.29/Modules/CMakeDetermineCompilerId.cmake:53 (__determine_compiler_id_test)
       >   /nix/store/yzi080r2c1zn2jzrhcfdv7dmr92yw07l-cmake-3.29.6/share/cmake-3.29/Modules/CMakeDetermineCUDACompiler.cmake:131 (CMAKE_DETERMINE_COMPILER_ID)
       >   CMakeLists.txt:74 (enable_language)
       >
       > 
       > -- Configuring incomplete, errors occurred!
       For full logs, run 'nix log /nix/store/ylwx6as34nfygmxz5lwjvcqb15p7xd5q-python3.12-bitsandbytes-0.43.3.drv'.
```

A fix is here: https://github.com/GaetanLepage/nixpkgs/compare/2af0c9ebf109911aebd324e6ae4a943ecfbebe6f...remexre:nixpkgs:bitsandbytes

**Comment by GaetanLepage** [comment]
> A fix is here: [GaetanLepage/nixpkgs@2af0c9e...remexre:nixpkgs:bitsandbytes](https://github.com/GaetanLepage/nixpkgs/compare/2af0c9ebf109911aebd324e6ae4a943ecfbebe6f...remexre:nixpkgs:bitsandbytes)

Thanks, the patch seems to work indeed.
I will wait for @SomeoneSerge's approval as he's much more knowledgeable than me on the CUDA stuff.

**Comment by GaetanLepage** [comment]
Well, even though it does build fine with `cudaSupport = true`, and the derivation is finalized successfully, it spits the following message:
```
Executing pythonImportsCheckPhase
Check whether the following modules can be imported: bitsandbytes
Could not load bitsandbytes native library: /nix/store/8x5i8xccb6h38iz22r8x3c4qg2pbrbg6-python3.12-bitsandbytes-0.44.1/lib/python3.12/site-packages/bitsandbytes/libbitsandbytes_cpu.so: cannot open shared object file: No such file or directory
Traceback (most recent call last):
  File "/nix/store/8x5i8xccb6h38iz22r8x3c4qg2pbrbg6-python3.12-bitsandbytes-0.44.1/lib/python3.12/site-packages/bitsandbytes/cextension.py", line 104, in <module>
    lib = get_native_library()
          ^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/8x5i8xccb6h38iz22r8x3c4qg2pbrbg6-python3.12-bitsandbytes-0.44.1/lib/python3.12/site-packages/bitsandbytes/cextension.py", line 91, in get_native_library
    dll = ct.cdll.LoadLibrary(str(binary_path))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/wfbjq35kxs6x83c3ncpfxdyl5gbhdx4h-python3-3.12.6/lib/python3.12/ctypes/__init__.py", line 460, in LoadLibrary
    return self._dlltype(name)
           ^^^^^^^^^^^^^^^^^^^
  File "/nix/store/wfbjq35kxs6x83c3ncpfxdyl5gbhdx4h-python3-3.12.6/lib/python3.12/ctypes/__init__.py", line 379, in __init__
    self._handle = _dlopen(self._name, mode)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: /nix/store/8x5i8xccb6h38iz22r8x3c4qg2pbrbg6-python3.12-bitsandbytes-0.44.1/lib/python3.12/site-packages/bitsandbytes/libbitsandbytes_cpu.so: cannot open shared object file: No such file or directory
```

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 343246`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python311Packages.bitsandbytes</li>
    <li>python311Packages.bitsandbytes.dist</li>
    <li>python312Packages.bitsandbytes</li>
    <li>python312Packages.bitsandbytes.dist</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python311Packages.bitsandbytes</li>
    <li>python311Packages.bitsandbytes.dist</li>
    <li>python312Packages.bitsandbytes</li>
    <li>python312Packages.bitsandbytes.dist</li>
  </ul>
</details>

---
### `x86_64-darwin`

---
### `aarch64-darwin`

**Review by samuela — APPROVED** [review]
diff lgtm

**Comment by samuela** [comment]
> OSError: /nix/store/8x5i8xccb6h38iz22r8x3c4qg2pbrbg6-python3.12-bitsandbytes-0.44.1/lib/python3.12/site-packages/bitsandbytes/libbitsandbytes_cpu.so: cannot open shared object file: No such file or directory

This looks like a job for `LD_DEBUG=libs`

**Comment by GaetanLepage** [comment]
Ok, I investigated some more.

## Problem
When `cudaSupport` is enabled, the cuda variant of the library is built `libbitsandbytes_cuda124.so` instead of the native CPU one `libbitsandbytes_cpu.so`.
However, at runtime, [they decide whether to load the cuda library or not depending on `torch.cuda.is_available()`](https://github.com/bitsandbytes-foundation/bitsandbytes/blob/9568735b21b9325e4789d6a5004517f2287f47c8/bitsandbytes/cuda_specs.py#L34-L35). However, this will return `False` on any system where there is no GPU available.
Hence, it falls back to trying to load the native `_cpu.so` library, which has not been built and thus doesn't exist.

**[Relevant code](https://github.com/bitsandbytes-foundation/bitsandbytes/blob/9568735b21b9325e4789d6a5004517f2287f47c8/bitsandbytes/cextension.py#L81-L90)**

In this function, `cuda_specs` will evaluate to `None` and the whole `if` statement will not trigger.

## Solution
The best solution I found was to:
1. ignore the `get_cuda_specs()` result and enter the `if` statement unconditionally.
2. bypass the call to `get_cuda_bnb_library_path(cuda_specs)` and hardcode the library path here (`PACKAGE_DIR / "libbitsandbytes_cuda124.so"`).

Now everything builds and imports fine. No more error messages.

**Comment by GaetanLepage** [comment]
> Ok, I investigated some more.
> ## Problem
> 
> When `cudaSupport` is enabled, the cuda variant of the library is built `libbitsandbytes_cuda124.so` instead of the native CPU one `libbitsandbytes_cpu.so`. However, at runtime, [they decide whether to load the cuda library or not depending on `torch.cuda.is_available()`](https://github.com/bitsandbytes-foundation/bitsandbytes/blob/9568735b21b9325e4789d6a5004517f2287f47c8/bitsandbytes/cuda_specs.py#L34-L35). However, this will return `False` on any system where there is no GPU available. Hence, it falls back to trying to load the native `_cpu.so` library, which has not been built and thus doesn't exist.
> 
> **[Relevant code](https://github.com/bitsandbytes-foundation/bitsandbytes/blob/9568735b21b9325e4789d6a5004517f2287f47c8/bitsandbytes/cextension.py#L81-L90)**
> 
> In this function, `cuda_specs` will evaluate to `None` and the whole `if` statement will not trigger.
> ## Solution
> 
> The best solution I found was to:
> 
>     1. ignore the `get_cuda_specs()` result and enter the `if` statement unconditionally.
> 
>     2. bypass the call to `get_cuda_bnb_library_path(cuda_specs)` and hardcode the library path here (`PACKAGE_DIR / "libbitsandbytes_cuda124.so"`).
> 
> 
> Now everything builds and imports fine. No more error messages.

## Result (with `cudaSupport = true`)

```
bitsandbytes> Running phase: pythonImportsCheckPhase
bitsandbytes> Executing pythonImportsCheckPhase
bitsandbytes> Check whether the following modules can be imported: bitsandbytes
┏━ Dependency Graph:
┃ ✔ python3.12-bitsandbytes-0.44.1 ⏱ 3m26s
┣━━━ Builds
┗━ ∑ ⏵ 0 │ ✔ 1 │ ⏸ 0 │ Finished at 12:21:35 after 3m29s
/nix/store/5n0nma25svsbw15x0qcj4sqyl8syrnyp-python3.12-bitsandbytes-0.44.1
```

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 343246`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python311Packages.bitsandbytes</li>
    <li>python311Packages.bitsandbytes.dist</li>
    <li>python312Packages.bitsandbytes</li>
    <li>python312Packages.bitsandbytes.dist</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python311Packages.bitsandbytes</li>
    <li>python311Packages.bitsandbytes.dist</li>
    <li>python312Packages.bitsandbytes</li>
    <li>python312Packages.bitsandbytes.dist</li>
  </ul>
</details>

---
### `x86_64-darwin`

---
### `aarch64-darwin`

File: pkgs/development/python-modules/bitsandbytes/default.nix
Comment (SomeoneSerge) [inline pkgs/development/python-modules/bitsandbytes/default.nix:40]: getDev, etc.

File: pkgs/development/python-modules/bitsandbytes/default.nix
Comment (SomeoneSerge) [inline pkgs/development/python-modules/bitsandbytes/default.nix:78]: Making sure: does it actually use the python package, or does it only need cmake the tool?

File: pkgs/development/python-modules/bitsandbytes/default.nix
Comment (SomeoneSerge) [inline pkgs/development/python-modules/bitsandbytes/default.nix:89]: cmaekSetupHook sets NVCC_APPEND_FLAGS too. Use `appendToVar` or `prependVar`

File: pkgs/development/python-modules/bitsandbytes/default.nix
Comment (SomeoneSerge) [inline pkgs/development/python-modules/bitsandbytes/default.nix:90]: This is normally done in `confgiurePhase` which `pkgs.cmake` in `nativeBuildInputs` would set automatically. Istead of specifying the `-D` flags in here, maybe move them to the nix attrset as `cmakeFlags` so the hook passes them on as a bash variable. If for some reason you choose not to use the hook but to call cmake yourself, reimplement the `flagsArray=(...) ; concatTo ...` functionality from the hook

File: pkgs/development/python-modules/bitsandbytes/default.nix
Comment (SomeoneSerge) [inline pkgs/development/python-modules/bitsandbytes/default.nix:19]: This will do for an interim solution but we'll be relying on upstream not changing the naming scheme. For now let's just add a `NOTE:` warning maybe, that when it breaks it's easier to find out what needs to be done

File: pkgs/development/python-modules/bitsandbytes/default.nix
Comment (SomeoneSerge) [inline pkgs/development/python-modules/bitsandbytes/default.nix:65]: Wow this really sounds like a bug in bitsanbytes, I'm surprised they're not reacting to the issue

Comment (SomeoneSerge) [reply pkgs/development/python-modules/bitsandbytes/default.nix:65]: FWIW on our side we could actually make their presumably-inteded behaviour work, only at double the cost: we could just make the cuda-version depend on the cpu-version.

The question is, is the `xxxxx_cuXXX.so` library usable in absence of GPUs?

Comment (GaetanLepage) [reply pkgs/development/python-modules/bitsandbytes/default.nix:65]: I think that this current approach is fine. If you set `cudaSupport = true`, it means that you want to use cuda.



> The question is, is the `xxxxx_cuXXX.so` library usable in absence of GPUs?

I don't think so.

File: pkgs/development/python-modules/bitsandbytes/default.nix
Comment (GaetanLepage) [reply pkgs/development/python-modules/bitsandbytes/default.nix:78]: It uses `python312Packages.cmake` but this is what every other python packages do as well.

File: pkgs/development/python-modules/bitsandbytes/default.nix
Comment (GaetanLepage) [reply pkgs/development/python-modules/bitsandbytes/default.nix:40]: Done.

File: pkgs/development/python-modules/bitsandbytes/default.nix
Comment (GaetanLepage) [reply pkgs/development/python-modules/bitsandbytes/default.nix]: I changed it to the same approach as in `mistral-rs`. Is it fine now ?

File: pkgs/development/python-modules/bitsandbytes/default.nix
Comment (GaetanLepage) [reply pkgs/development/python-modules/bitsandbytes/default.nix]: I was able to successfully rely on the cmake hook.
I still manually run `make` in the `preBuild` phase.

File: pkgs/development/python-modules/bitsandbytes/default.nix
Comment (GaetanLepage) [reply pkgs/development/python-modules/bitsandbytes/default.nix:19]: Done, I have just added a note.

File: pkgs/development/python-modules/bitsandbytes/default.nix
Comment (samuela) [inline pkgs/development/python-modules/bitsandbytes/default.nix:68]:
```suggestion
```

Comment (GaetanLepage) [reply pkgs/development/python-modules/bitsandbytes/default.nix:68]: It was on purpose to make this long comment more readable. I can remove it if necessary.

Comment (samuela) [reply pkgs/development/python-modules/bitsandbytes/default.nix:68]: oh oops, github's diff viewer showed there being an empty line between this and line 69. i misunderstood -- sorry for the noise!

File: pkgs/development/python-modules/bitsandbytes/default.nix
Comment (SomeoneSerge) [reply pkgs/development/python-modules/bitsandbytes/default.nix:66]: > I think that this current approach is fine. If you set cudaSupport = true, it means that you want to use cuda.

Well the upstream has implemented this (broken?) dynamic dispatching logic so I guess they intended it the other way. Anyway, implementing that logic from scratch is not in-scope here

File: pkgs/development/python-modules/bitsandbytes/default.nix
Comment (SomeoneSerge) [inline pkgs/development/python-modules/bitsandbytes/default.nix:31]: drop the last? Btw `getDev libcusparse` has `.lib` in `propagatedBuildInputs`, I forget if that matters to `symlinkJoin`

File: pkgs/development/python-modules/bitsandbytes/default.nix
Comment (SomeoneSerge) [inline pkgs/development/python-modules/bitsandbytes/default.nix:92]: Drop `-D`, it's already prepended by `cmakeFeature`

**Review by SomeoneSerge — APPROVED** [review]

File: pkgs/development/python-modules/bitsandbytes/default.nix
Comment (GaetanLepage) [reply pkgs/development/python-modules/bitsandbytes/default.nix:66]: > Well the upstream has implemented this (broken?) dynamic dispatching logic so I guess they intended it the other way. Anyway, implementing that logic from scratch is not in-scope here

Yes, indeed.

File: pkgs/development/python-modules/bitsandbytes/default.nix
Comment (GaetanLepage) [reply pkgs/development/python-modules/bitsandbytes/default.nix:31]: Oh, well the last line `libcusparse.lib` is clearly a mistake and is not needing because I have added `lib.getLib libcusparse` above.
Now, having said that, you suggest also getting rid of that and only keeping `lib.getDev libcusparse` right ?

Comment (GaetanLepage) [reply pkgs/development/python-modules/bitsandbytes/default.nix:31]: It works with `lib.getDev libcusparse` alone.

**Comment by GaetanLepage** [comment]
Ok I was able to make it build in the end.
However, I am not sure that the `cuda-native-redist` and `cuda-common-redist` are as lean as they can be.
Can you double check my changes @SomeoneSerge please ?

**Comment by SomeoneSerge** [comment]
> However, I am not sure that the cuda-native-redist and cuda-common-redist are as lean as they can be.

That matters fairly little when using `symlinkJoin`: the closure will be huge regardless.

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 343246`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python311Packages.bitsandbytes</li>
    <li>python311Packages.bitsandbytes.dist</li>
    <li>python312Packages.bitsandbytes</li>
    <li>python312Packages.bitsandbytes.dist</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python311Packages.bitsandbytes</li>
    <li>python311Packages.bitsandbytes.dist</li>
    <li>python312Packages.bitsandbytes</li>
    <li>python312Packages.bitsandbytes.dist</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:fast_forward: 4 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.bitsandbytes</li>
    <li>python311Packages.bitsandbytes.dist</li>
    <li>python312Packages.bitsandbytes</li>
    <li>python312Packages.bitsandbytes.dist</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:fast_forward: 4 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.bitsandbytes</li>
    <li>python311Packages.bitsandbytes.dist</li>
    <li>python312Packages.bitsandbytes</li>
    <li>python312Packages.bitsandbytes.dist</li>
  </ul>
</details>

**Pull request merged**
